### ISSUE TITLE
Allowing DateFormat in the parser

### ISSUE THREAD

**Issue by venkatd** [petitparser/dart-petitparser#84]
In our application, depending on the locale of the user, we have a DateFormat that varies by the user (intl library). For example:

```dart
final kMessageDateTime = DateFormat('E MMM d, h:mm a');
```

I was considering to build actual grammars from these format tokens but that feels like overkill as there could be a bunch depending on the country of the user.

How might I implement something like the following?
```
final localDateTime = formattedDateTime('E MMM d, h:mm a');
final dates = localDateTime.matchesSkipping(someMessageWithTimeReferences);
```

From looking through the source, it seems like I would need to extend `Parser` and expose a helper method. If that's the case, any ideas on where I can begin? Any good examples I can look at?

Somewhat related... I see some parsers implement a `fastParseOn` and a `parseOn`. Why is `parseOn` necessary if `fastParseOn` is implemented?

**Comment by venkatd** [comment]
I'm torn between generating a grammar based on a DateFormat object vs. creating a custom parser.

For example, the following seems like it could work.


```dart
// mostly hardcoded for just "E MMM d, h:mm a" to test things out
Parser<String> dateTime(DateFormat format) {
  final shortMonth = anyWordIgnoringCase(format.dateSymbols.SHORTMONTHS);
  final dateMday = (num2Digit(min: 10, max: 31) | pattern('1-9')).flatten();
  final hour12 = (num2Digit(min: 1, max: 12) | pattern('0-9')).flatten();
  final amPm = anyWordIgnoringCase(['am', 'pm']);
  final timeMinute = num2Digit(max: 59);
  return (shortMonth &
          char(' ') &
          dateMday &
          string(', ') &
          hour12 &
          char(':') &
          timeMinute &
          char(' ') &
          amPm)
      .flatten();
}
```

It looks like I could read from `DateFormat.dateSymbols` and map each format element (E MMM, d, h:mm, a, etc) to a corresponding parser. But it seems like a lot of code and unsure if it's cleaner to write a parser.

**Comment by renggli** [comment]
You could create a parser that builds another parser relatively easy:

```dart
final day = 'dd'.toParser().map((token) => digit()
    .repeat(2)
    .flatten()
    .map((value) => MapEntry(#day, int.parse(value))));
final month = 'mm'.toParser().map((token) => digit()
    .repeat(2)
    .flatten()
    .map((value) => MapEntry(#month, int.parse(value))));
final year = 'yyyy'.toParser().map((token) => digit()
    .repeat(4)
    .flatten()
    .map((value) => MapEntry(#year, int.parse(value))));

final spacing = whitespace().map((token) =>
    whitespace().star().map((value) => const MapEntry(#unused, 0)));
final verbatim = any().map(
    (token) => token.toParser().map((value) => const MapEntry(#unused, 0)));

final entries = [day, month, year, spacing, verbatim].toChoiceParser();
final format = entries
    .star()
    .end()
    .map((parsers) => parsers.toSequenceParser().map((entries) {
          final arguments = Map.fromEntries(entries);
          return DateTime(
            arguments[#year] ?? DateTime.now().year,
            arguments[#month] ?? DateTime.january,
            arguments[#day] ?? 1,
          );
        }));
```

Then you can create date formats like so: 

```dart
final iso = format.parse('yyyy-mm-dd').value;
final date = iso.parse('1980-06-11').value;
print(date);
````

I do not really see why you want to subclass your own `Parser`?  `fastParseOn` is an optimized version of `parseOn`, that does not produce a result, but can provide a significant speed boost in some cases. Normally you should only need to implement `parseOn`.

**Comment by venkatd** [comment]
Thanks, I tweaked the code a bit and will post what I come up with later. I think because I have `implicit-casts: false` enabled in my `analysis_options.yml`, I had to add some explicit casting `cast` and `castList`. Otherwise the functions are returning `List<dynamic>`

I think if the extensions had generics added, these casting steps wouldn't have to be added. Is this something that's worth improving? I could create a separate issue to track if it seems worth while.

**Comment by renggli** [comment]
Yeah, I already noticed myself that the choice parsers was missing a generic type. Fixed this in b5e85d6e55fcebdf198adb5b169b122c6b332d68. I was on the [nullsafety](https://github.com/petitparser/dart-petitparser/tree/nullsafety) branch, so that might explain some of the other differences. If you have any other suggestions of improving the typing feel free to add bugs/pull-requests.
